gen/cpp/sym/util/type_ops.h

```diff
@@ -36,6 +37,280 @@ template <typename T>
 static constexpr const bool kIsSparseEigenType =
     std::is_base_of<Eigen::SparseMatrix<typename T::Scalar>, T>::value;
 
+inline constexpr bool IsEigenType(const type_t type) {
+  switch (type.value) {
+    case type_t::VECTOR1:
+    case type_t::VECTOR2:
+    case type_t::VECTOR3:
+    case type_t::VECTOR4:
+    case type_t::VECTOR5:
+    case type_t::VECTOR6:
+    case type_t::VECTOR7:
+    case type_t::VECTOR8:
+    case type_t::VECTOR9:
+    case type_t::MATRIX12:
+    case type_t::MATRIX13:
+    case type_t::MATRIX14:
+    case type_t::MATRIX15:
+    case type_t::MATRIX16:
+    case type_t::MATRIX17:
+    case type_t::MATRIX18:
+    case type_t::MATRIX19:
+    case type_t::MATRIX22:
+    case type_t::MATRIX23:
+    case type_t::MATRIX24:
+    case type_t::MATRIX25:
+    case type_t::MATRIX26:
+    case type_t::MATRIX27:
+    case type_t::MATRIX28:
+    case type_t::MATRIX29:
+    case type_t::MATRIX32:
+    case type_t::MATRIX33:
+    case type_t::MATRIX34:
+    case type_t::MATRIX35:
+    case type_t::MATRIX36:
+    case type_t::MATRIX37:
+    case type_t::MATRIX38:
+    case type_t::MATRIX39:
+    case type_t::MATRIX42:
+    case type_t::MATRIX43:
+    case type_t::MATRIX44:
+    case type_t::MATRIX45:
+    case type_t::MATRIX46:
+    case type_t::MATRIX47:
+    case type_t::MATRIX48:
+    case type_t::MATRIX49:
+    case type_t::MATRIX52:
+    case type_t::MATRIX53:
+    case type_t::MATRIX54:
+    case type_t::MATRIX55:
+    case type_t::MATRIX56:
+    case type_t::MATRIX57:
+    case type_t::MATRIX58:
+    case type_t::MATRIX59:
+    case type_t::MATRIX62:
+    case type_t::MATRIX63:
+    case type_t::MATRIX64:
+    case type_t::MATRIX65:
+    case type_t::MATRIX66:
+    case type_t::MATRIX67:
+    case type_t::MATRIX68:
+    case type_t::MATRIX69:
+    case type_t::MATRIX72:
+    case type_t::MATRIX73:
+    case type_t::MATRIX74:
+    case type_t::MATRIX75:
+    case type_t::MATRIX76:
+    case type_t::MATRIX77:
+    case type_t::MATRIX78:
+    case type_t::MATRIX79:
+    case type_t::MATRIX82:
+    case type_t::MATRIX83:
+    case type_t::MATRIX84:
+    case type_t::MATRIX85:
+    case type_t::MATRIX86:
+    case type_t::MATRIX87:
+    case type_t::MATRIX88:
+    case type_t::MATRIX89:
+    case type_t::MATRIX92:
+    case type_t::MATRIX93:
+    case type_t::MATRIX94:
+    case type_t::MATRIX95:
+    case type_t::MATRIX96:
+    case type_t::MATRIX97:
+    case type_t::MATRIX98:
+    case type_t::MATRIX99:
+      return true;
+    case type_t::ROT2:
+    case type_t::ROT3:
+    case type_t::POSE2:
+    case type_t::POSE3:
+    case type_t::UNIT3:
+    case type_t::SCALAR:
+    case type_t::ATAN_CAMERA_CAL:
+    case type_t::DOUBLE_SPHERE_CAMERA_CAL:
+    case type_t::EQUIRECTANGULAR_CAMERA_CAL:
+    case type_t::LINEAR_CAMERA_CAL:
+    case type_t::POLYNOMIAL_CAMERA_CAL:
+    case type_t::SPHERICAL_CAMERA_CAL:
+      return false;
+    default:
+      SYM_ASSERT(false);
```

Comment:
We have `-Wall` and `-Werror` turned on right? If so, maybe it would be better to not have the default case so that we'll get a compiler error if we forget a case (rather than having to rely on a runtime error).

An example error:
```
<source>:10:12: error: enumeration value 'B' not handled in switch [-Werror=switch]
   10 |     switch (t) {
      |            ^
cc1plus: all warnings being treated as errors
```

[godbolt example that demos this](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:1,endLineNumber:12,positionColumn:1,positionLineNumber:12,selectionStartColumn:1,selectionStartLineNumber:12,startColumn:1,startLineNumber:12),source:'%23include+%3Ciostream%3E%0A%23include+%3Cmemory%3E%0A%0Aenum+type+%7B%0A++++A,%0A++++B%0A%7D%3B%0A%0Achar+TypeToChar(type+t)+%7B%0A++++switch+(t)+%7B%0A++++++++case+type::A:%0A++++++++++++return+!'a!'%3B%0A++++%7D%0A++++return+!'0!'%3B%0A%7D%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:g132,compilerName:'',compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'--std%3Dc%2B%2B14+-Wall+-Werror',overrides:!(),source:1,stdinPanelShown:'1',wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+gcc+13.2+(C%2B%2B,+Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Reply:
I don't think we can here because we've split the switch to just handle the matrix types here and the other ones in the other function, so neither of them actually handles all of type_t

Reply:
Ah I see this one actually does include all of them, yeah I'll do that

Reply:
Done

Reply:
gcc (5 and 11) can't seem to figure out this switch is exhaustive and flow out the bottom is impossible (or, I guess maybe gcc doesn't want to assume the enum holds a valid value?), so need an assert there.  gcc 5 also doesn't like the `throw` in a constexpr function, but we aren't using these in a constexpr context anyway so just removed that